### ISSUE TITLE
Prefer package.json5 over package.json

### ISSUE THREAD

**Issue by JasonKleban** [pnpm/pnpm#3027]
### Code to reproduce the issue:

`"preinstall": "npx only-allow pnpm"` documented at https://pnpm.js.org/en/only-allow-pnpm is awesome!  However it seems impossible to use it in conjunction with `package.json5`'s file extension and its lovely support for comments AND protect against package managers such as yarn which do not recognize .json5.

I thought I'd be clever and create a package.json file that has ONLY the following (in addition to the same script in my package.json5):

```
{
  "scripts": {
    "preinstall": "npx only-allow pnpm"
  }
}
```

But unfortunately, pnpm seems to prefer package.json over package.json5.  Also, as appropriate, pnpm errors on comments in .json files, so we can't just pretend it's json5/jsonc.

### Expected behavior:

pnpm should prefer package.json5 (and yaml) over package.json as the newer and more capable formats.

### Actual behavior:

pnpm seems to prefer package.json over package.json5 (and yaml)

**Comment by JasonKleban** [comment]
I guess the order change would be done in these two places?

https://github.com/pnpm/pnpm/blob/f9befbfb534e4027fd664b335eaf478ee6d890ea/packages/read-project-manifest/src/index.ts#L60-L89

https://github.com/pnpm/pnpm/blob/f9befbfb534e4027fd664b335eaf478ee6d890ea/packages/find-packages/src/index.ts#L76-L81

**Comment by JasonKleban** [comment]
Ooooh, this seems like a blocker https://github.com/pnpm/pnpm/issues/2008 before this precedence would matter much.

**Comment by zkochan** [comment]
I guess I don't have objections to change the order but I am also not sure it was a good idea to add json5/yaml support. I received a lot of criticism on Twitter for that. cc @pnpm/collaborators

**Comment by juanpicado** [comment]
Let's imagine you have `package.yaml` or `package.json5` with comments, on `pnpm publish` what exactly goes to the registry?

**Comment by zkochan** [comment]
pnpm converts those files to regular json files before publish

**Comment by juanpicado** [comment]
> I received a lot of criticism on Twitter for that. cc

What was exactly the thoughts other shared?

**Comment by zkochan** [comment]
https://twitter.com/bitandbang/status/1124993994881155072

**Comment by juanpicado** [comment]
> https://twitter.com/bitandbang/status/1124993994881155072

I see, good point https://nodejs.org/api/packages.html

**Comment by nikoladev** [comment]
I'd vote for not doing anything that can break workflows. So could this maybe be solved with an opt-in CLI flag?

**Comment by aminya** [comment]
I recommend JSOX/JSON6:

https://github.com/d3x0r/sack.vfs

**Comment by JasonKleban** [comment]
Whatever new format are chosen for support, the benefits of human-readability crucially include ordered-member and trivia (whitespace and comments) preservation when the package file is subjected to automated changes (like `pnpm add`).

I imagine that the simplest way to support this is for the package file updater to _merge_ the newly calculated javascript object data _into_ the original file as a final step - and this will require a parser library with support for merging ASTs with trivia.  Are there any next gen notation formats that can provide this?  I don't see these features in json5, json6, jsox, jsonc.  [hjson has `-rt`/`keepWsc` for retaining comments and whitespace](https://github.com/hjson/hjson-js#modify--keep-comments) - but based on how it seems to work, I guess that it cannot preserve object member order?

**Comment by bdchauvette** [comment]
> I am also not sure it was a good idea to add json5/yaml support

IMO it's nice to have in theory, but it's not really useful in practice if you need or want to:

- use dev tools that only look for `package.json`, e.g. [`husky@<5`](https://github.com/typicode/husky/blob/cbb0af7cb5ae30380809c9d11f93902e98fc2d30/src/installer/bin.ts#L61), [`cosmicconfig`](https://www.npmjs.com/package/cosmiconfig) for non-rc files, [`eslint-plugin-import`](https://github.com/benmosher/eslint-plugin-import/blob/8c1a65d0bdf446b03e513cc2b7377ca49232a6c3/src/rules/no-extraneous-dependencies.js#L71)
- export metadata from your own `package.json`, e.g. `exports.version = require('./package.json').version` (which is admittedly probably not a best practice, but :shrug:)
- use [`"type": "module"`](https://nodejs.org/api/packages.html#packages_type) to write native ESM modules without having to use `.mjs`
- use [`"imports"`](https://nodejs.org/api/packages.html#packages_imports) to get native import aliases

There are probably a lot of people -- maybe even most -- that are fine with the tradeoffs, but I think the presence of a `package.json` file is such a fundamental assumption in the Node ecosystem that it doesn't really seem worthwhile (IMO) to support non-json configs in the first place.

However, removing support would also be a large breaking change for people that rely on it, so if `pnpm` were to remove support for non-json configs, it should probably be done gradually, e.g. no change in v5, deprecate in v6, actually remove in v7.

**Comment by JasonKleban** [comment]
@bdchauvette the world as to keep turning somehow!  If pnpm is the near future of package management, then it seems that support for a new format would have to start here, no?

**Comment by DrSensor** [comment]
Fyi, most of the tools (tsc, esbuild, ...) that read `tsconfig.json` support comment and trailing comma. It's identified as JSON with comment (`jsonc`) If you open it with VSCode. Maybe it's good idea to borrow that idea 🙂

**Comment by vjpr** [comment]
I think its worthy of a config setting.

When using `package.json5`, since many tools rely on `package.json`, one always needs to be present. It would need to be synced.

We should add a hook called `writePackage` to `pnpmfile.js` to allow syncing when running `pnpm add|remove|link`. See: https://github.com/pnpm/pnpm/discussions/5624

**Comment by rlidwka** [comment]
I agree with this change for the reasons I've explained in https://github.com/pnpm/pnpm/issues/5541#issuecomment-1853105895 (which seems to be a duplicate issue).

> I am also not sure it was a good idea to add json5/yaml support.

Some would agree with this, others wouldn't. Personally, I've forked `npm` back in 2014 because of *this exact issue* (but it didn't catch on at the time), so I'm certainly thankful that I can use json5 here.

**Comment by bmulholland** [comment]
We're now using package.json5 pretty nicely with pnpm. We have both a package.json5 and a package.json (both in git), edit the json5 only, and compile the json5 to json with any changes. The following scripts entries do that compilation anytime we do a `pnpm install`, so we don't forget:
```
    // Enable comments in our packgae spec with json5
    // Echo a newline to ensure the last line is always a newline, otherwise prettier will complain
    // See https://github.com/nodejs/node/issues/40714
    "compile-package-json": "json5 -o package.json --space 2 package.json5 && echo '' >> package.json",
    // Update the package.json from the json5 before any install action
    // https://pnpm.io/scripts#pnpmdevpreinstall
    "pnpm:devPreinstall": "if [ -d node_modules ] && [ -f node_modules/.bin/json5 ]; then pnpm compile-package-json; else echo 'json5 is not installed or node_modules is missing, skipping compile-package-json'; fi",
```

**Comment by neochief** [comment]
@bmulholland's solution would require changes in the context of the monorepo. And even after, you're still stuck with problems described above: you need to run all of the install scripts twice.

**Comment by bmulholland** [comment]
Yeah, I was wrong above :(. The problem is that pnpm doesn't reload package.json after the devPreinstall. So the compile happens, but isn't picked up until later. So after any changes to package.json5, you either need to run install twice, or just `pnpm compile-package-json && pnpm install`.

**Comment by Flimm** [comment]
If both `package.json` and `package.json5` exist, how should pnpm behave? Here are some options:

1. It can use `package.json` and ignore `package.json5` (current behaviour)
2. It can use `package.json5` and ignore `package.json` (the proposal of this GitHub issue)
3. It can check that both files compile to the equivalent value, and throw an error if they are not equivalent
4. It can throw an error, requiring the developer to delete one of the two files

Solutions 1 or 2 could also emit a warning if the two files are not equivalent.

Personally, I would prefer solution 4 or solution 3, to avoid unexpected behaviour if the two files get out of sync with each other.

**Comment by EfeDurmaz16** [comment]
Triage note: the still-open draft PR pnpm/pnpm#11372 proposes the `preferredManifestFormat` setting for this use case. It also includes related workspace manifest deduplication work.

I am leaving this issue open and avoiding a competing implementation while that proposal is unresolved.

Written by an agent (Codex, GPT-6).

**Comment by bjones1** [comment]
The title of this issue and much of its discussion is focused on preferring `package.json5` over `package.json`. However, #15129 states "pnpm 12.5.1 ignores `package.json5`," which is the bug I see in practice. Should a separate issue therefore be opened for the issue "pnpm 12.5.1 ignores `package.json5`," or are the devs (thanks, @zkochan, for making this a priority!) planning on a fix that will address both problems?

Thanks again for the hard work the devs have put in to make this a great project!

**Comment by simon-aira** [comment]
I'd be happy to (attempt to) clean up and un-draft my PR #11372 if you agree that the mechanism I propose there makes sense!

**Comment by EfeDurmaz16** [comment]
The approach makes sense to me
